File: tiq-e/array/6/main.py
from time import perf_counter
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res = []
        dictionary = {}
        for n in nums1:
            if n in dictionary:
                dictionary[n] += 1
            else:
                dictionary[n] = 1
        for n in nums2:
            if n in dictionary:
                res.append(n)
                dictionary[n] -= 1
            if n in dictionary and dictionary[n] == 0:
                del dictionary[n]
        return res


arr = [4, 1, 2, 1, 2]
arr2 = [3, 1, 2, 1, 2]

start = perf_counter()
result = Solution().intersect(arr, arr2)
end = perf_counter()
print("Duration: " + str(end - start))
print("Result: " + str(result))

File: tiq-e/array/6/main.go
package main

import (
	"fmt"
	"time"
)

func intersect(nums1 []int, nums2 []int) []int {
	cache := make(map[int]int)
	for _, v := range nums1 {
		cache[v]++
	}

	var res []int
	for _, v := range nums2 {
		if count, ok := cache[v]; ok {
			if count == 0 {
				continue
			}
			cache[v]--
			res = append(res, v)
		}
	}
	return res
}

func main() {
	arr := []int{3, 2, 2, 6, 4}
	arr2 := []int{3, 2, 2, 6, 3}

	var start = time.Now()
	var result = intersect(arr, arr2)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}
